File: challenge-1/submissions/RezaSi/app/main.py
"""Reference URL shortener — FastAPI + Redis + MongoDB.

Design:

- **MongoDB** is the source of truth. Documents look like
    {"_id": <code>, "url": <url>, "hits": <int>}
  with a unique index on `url` so duplicate inserts are idempotent.

- **Redis** is two things at once:
    1. A read-through cache for `code -> url` lookups. Resolves never
       touch Mongo on a cache hit.
    2. A transient counter `h:<code>` that buffers hits between flushes.
       A background task moves them to Mongo every ~1s. That keeps the
       hot redirect path off the database.

- **Code generation** is deterministic: `base62(sha256(url)[:8])`. That
  makes `POST /shorten` idempotent without a read-before-write. The 64-bit
  prefix gives ≈1.8×10^19 distinct codes; the birthday-paradox collision
  probability is negligible at any realistic scale.

What makes this persist a `docker compose restart`:

- Mongo writes go to /data/db, which is the `mongodata` named volume.
- Restarting the app container has zero effect on Mongo's data.
- Redis state is intentionally volatile — losing the cache means slower
  reads for a few seconds, not data loss.
"""

from __future__ import annotations

import asyncio
import hashlib
import logging
import os
import string
from contextlib import asynccontextmanager
from typing import Optional

import redis.asyncio as aioredis
from fastapi import FastAPI, Request
from fastapi.responses import JSONResponse, RedirectResponse
from motor.motor_asyncio import AsyncIOMotorClient
from pymongo.errors import DuplicateKeyError


logger = logging.getLogger("shortener")

MONGO_URL = os.environ.get("MONGO_URL", "mongodb://mongo:27017")
REDIS_URL = os.environ.get("REDIS_URL", "redis://cache:6379")
DB_NAME = os.environ.get("MONGO_DB", "shortener")
COLL_NAME = os.environ.get("MONGO_COLL", "codes")

CODE_ALPHABET = string.ascii_letters + string.digits  # 62 chars
CODE_LENGTH = 7
HITS_KEY_PREFIX = "h:"
CACHE_KEY_PREFIX = "u:"
CACHE_TTL_SECONDS = 60 * 60  # 1h, plenty for a 60s load test
FLUSH_INTERVAL_SECONDS = 1.0


def _base62(num: int, length: int) -> str:
    base = len(CODE_ALPHABET)
    chars = []
    for _ in range(length):
        chars.append(CODE_ALPHABET[num % base])
        num //= base
    return "".join(reversed(chars))


def make_code(url: str) -> str:
    """Deterministic, idempotent short code for a URL.

    sha256 → first 8 bytes → big-endian uint64 → base62 truncated to 7 chars.
    """
    digest = hashlib.sha256(url.encode("utf-8")).digest()
    return _base62(int.from_bytes(digest[:8], "big"), CODE_LENGTH)


# ---------------------------------------------------------------------------
# Lifespan: wire Mongo + Redis, ensure indices, start background flusher.
# ---------------------------------------------------------------------------


async def _flush_loop(app: FastAPI) -> None:
    """Periodically move buffered hit counts from Redis to Mongo.

    We use `GETDEL` (atomic get-and-delete) on each `h:<code>` key so we
    don't race with the resolve path. Failures are logged and the loop
    continues.
    """
    redis: aioredis.Redis = app.state.redis
    coll = app.state.coll
    while True:
        try:
            await asyncio.sleep(FLUSH_INTERVAL_SECONDS)
            cursor = 0
            updates: dict[str, int] = {}
            while True:
                cursor, keys = await redis.scan(
                    cursor=cursor, match=f"{HITS_KEY_PREFIX}*", count=200
                )
                for raw in keys:
                    key = raw if isinstance(raw, str) else raw.decode()
                    code = key[len(HITS_KEY_PREFIX) :]
                    val = await redis.getdel(key)
                    if val is None:
                        continue
                    try:
                        n = int(val)
                    except (TypeError, ValueError):
                        continue
                    if n > 0:
                        updates[code] = updates.get(code, 0) + n
                if cursor == 0:
                    break
            for code, n in updates.items():
                try:
                    await coll.update_one({"_id": code}, {"$inc": {"hits": n}})
                except Exception:
                    logger.exception("failed to flush %d hits for %s", n, code)
        except asyncio.CancelledError:
            raise
        except Exception:
            logger.exception("flush loop iteration failed")


async def _ensure_ready(app: FastAPI) -> None:
    """Background task: poll Mongo + Redis until both answer, then mark ready.

    We intentionally do NOT block lifespan startup on this. uvicorn binds
    port 8080 immediately, /healthz returns 503 until both backends are up.
    That keeps the grader's health probe useful even while a tight
    container is still warming Mongo / Redis from cold.
    """
    mongo = app.state.mongo
    redis = app.state.redis
    coll = app.state.coll
    while not app.state.shutdown:
        try:
            await mongo.admin.command("ping")
            await redis.ping()
            await coll.create_index("url", unique=True)
            app.state.ready = True
            return
        except Exception:
            await asyncio.sleep(0.5)


@asynccontextmanager
async def lifespan(app: FastAPI):
    # Non-blocking startup: we wire the clients here but never await on
    # them. The clients are lazy — actual TCP happens on first call.
    mongo = AsyncIOMotorClient(MONGO_URL, serverSelectionTimeoutMS=2000)
    coll = mongo[DB_NAME][COLL_NAME]
    redis = aioredis.from_url(REDIS_URL, decode_responses=True)

    app.state.mongo = mongo
    app.state.coll = coll
    app.state.redis = redis
    app.state.ready = False
    app.state.shutdown = False

    ready_task = asyncio.create_task(_ensure_ready(app))
    flush_task = asyncio.create_task(_flush_loop(app))

    try:
        yield
    finally:
        app.state.shutdown = True
        for task in (ready_task, flush_task):
            task.cancel()
            try:
                await task
            except (asyncio.CancelledError, Exception):
                pass
        await redis.aclose()
        mongo.close()


app = FastAPI(lifespan=lifespan)


def _short_url(request: Request, code: str) -> str:
    base = str(request.base_url).rstrip("/")
    return f"{base}/{code}"


# ---------------------------------------------------------------------------
# Endpoints
# ---------------------------------------------------------------------------


@app.get("/healthz")
async def healthz() -> JSONResponse:
    """200 only when both Mongo and Redis have answered at least once.

    `app.state.ready` is flipped by the background `_ensure_ready` task
    that runs alongside the server. Returning 503 while still warming up
    is what lets uvicorn bind port 8080 immediately and still give the
    grader an honest "not ready yet" signal.
    """
    if not getattr(app.state, "ready", False):
        return JSONResponse({"status": "starting"}, status_code=503)
    try:
        await app.state.redis.ping()
        await app.state.mongo.admin.command("ping")
    except Exception:
        return JSONResponse({"status": "degraded"}, status_code=503)
    return JSONResponse({"status": "ok"})


@app.post("/shorten")
async def shorten(request: Request):
    body_bytes = await request.body()
    try:
        # Avoid request.json() so we control malformed-JSON errors.
        import json as _json
        body = _json.loads(body_bytes) if body_bytes else None
    except ValueError:
        return JSONResponse({"error": "invalid JSON"}, status_code=400)
    if not isinstance(body, dict):
        return JSONResponse({"error": "body must be a JSON object"}, status_code=400)

    url = body.get("url")
    if not isinstance(url, str) or not url:
        return JSONResponse({"error": "url is required"}, status_code=400)
    if not (url.startswith("http://") or url.startswith("https://")):
        return JSONResponse({"error": "url must use http or https"}, status_code=400)

    code = make_code(url)
    coll = app.state.coll
    redis = app.state.redis

    # Try to insert; if the URL is already present, return the existing code.
    try:
        await coll.insert_one({"_id": code, "url": url, "hits": 0})
        await redis.set(f"{CACHE_KEY_PREFIX}{code}", url, ex=CACHE_TTL_SECONDS)
        return JSONResponse(
            {"code": code, "short_url": _short_url(request, code)},
            status_code=201,
        )
    except DuplicateKeyError:
        # `_id` collision OR `url` unique-index collision. Look up the
        # existing code for this URL and return it.
        doc = await coll.find_one({"url": url})
        if doc is None:
            # Race lost: someone deleted between the insert error and the
            # lookup. Retry the insert by recursing once.
            await coll.insert_one({"_id": code, "url": url, "hits": 0})
            existing = code
        else:
            existing = doc["_id"]
        return JSONResponse(
            {"code": existing, "short_url": _short_url(request, existing)},
            status_code=200,
        )


@app.get("/api/codes/{code}")
async def metadata(code: str):
    coll = app.state.coll
    redis = app.state.redis
    doc = await coll.find_one({"_id": code})
    if doc is None:
        return JSONResponse({"error": "not found"}, status_code=404)
    pending = await redis.get(f"{HITS_KEY_PREFIX}{code}")
    try:
        pending_int = int(pending) if pending is not None else 0
    except (TypeError, ValueError):
        pending_int = 0
    return JSONResponse(
        {
            "code": code,
            "url": doc["url"],
            "hits": int(doc.get("hits", 0)) + pending_int,
        }
    )


@app.get("/{code}")
async def resolve(code: str):
    # Reject obvious non-codes (anything with a slash or longer than 16 chars).
    # FastAPI's path matcher won't match strings with slashes anyway, but
    # being explicit makes the failure modes clearer.
    if len(code) < 4 or len(code) > 16 or not code.isalnum():
        return JSONResponse({"error": "not found"}, status_code=404)

    redis = app.state.redis
    cache_key = f"{CACHE_KEY_PREFIX}{code}"
    url: Optional[str] = await redis.get(cache_key)
    if url is None:
        coll = app.state.coll
        doc = await coll.find_one({"_id": code}, {"url": 1})
        if doc is None:
            return JSONResponse({"error": "not found"}, status_code=404)
        url = doc["url"]
        # Cache for subsequent reads; missing keys are not cached.
        await redis.set(cache_key, url, ex=CACHE_TTL_SECONDS)

    # Fire-and-forget hit increment in Redis. The background flusher
    # persists this to Mongo within ~1s, well inside the spec's slack.
    try:
        await redis.incr(f"{HITS_KEY_PREFIX}{code}")
    except Exception:
        # If Redis hiccups, the redirect itself must still succeed.
        pass

    return RedirectResponse(url=url, status_code=302)

File: challenge-1/submissions/RezaSi/main.go
// RezaSi reference submission for challenge-1 (URL Shortener).
//
// Two containers: this Go service and a Redis with AOF persistence on a
// named volume. Redis is the only source of truth — URL mappings and the
// hit counter both live there. AOF (`appendonly yes`, `appendfsync
// everysec`) carries data across `docker compose restart` with at most
// one second of write loss.
//
// Codes are 7 random base62 characters (~41 bits of entropy). The spec
// no longer requires same-URL-same-code, so each POST mints a fresh
// code; on the (vanishingly rare) collision with an existing key, we
// re-roll up to a handful of times. No reverse url->code index needed.
package main

import (
	"context"
	"crypto/rand"
	"encoding/json"
	"errors"
	"io"
	"log"
	"net/http"
	"os"
	"os/signal"
	"regexp"
	"strconv"
	"strings"
	"sync/atomic"
	"syscall"
	"time"

	"github.com/redis/go-redis/v9"
)

const (
	keyURLPrefix = "u:"
	keyHitPrefix = "h:"
	codeLen      = 7
	maxMintTries = 8
)

var (
	base62Alphabet = []byte("0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz")
	codeRegex      = regexp.MustCompile(`^[A-Za-z0-9]{4,16}$`)
)

type server struct {
	rdb        *redis.Client
	publicHost string
	ready      atomic.Bool
}

func main() {
	addr := getenv("REDIS_ADDR", "cache:6379")
	publicHost := strings.TrimRight(getenv("PUBLIC_HOST", "http://localhost:8080"), "/")

	rdb := redis.NewClient(&redis.Options{
		Addr:         addr,
		PoolSize:     64,
		MinIdleConns: 8,
		DialTimeout:  2 * time.Second,
		ReadTimeout:  1 * time.Second,
		WriteTimeout: 1 * time.Second,
		MaxRetries:   2,
	})
	s := &server{rdb: rdb, publicHost: publicHost}
	go s.waitForRedis()

	mux := http.NewServeMux()
	mux.HandleFunc("GET /healthz", s.healthz)
	mux.HandleFunc("POST /shorten", s.shorten)
	mux.HandleFunc("GET /api/codes/{code}", s.metadata)
	mux.HandleFunc("GET /{code}", s.resolve)

	httpServer := &http.Server{
		Addr:              ":8080",
		Handler:           mux,
		ReadHeaderTimeout: 3 * time.Second,
	}

	sigCh := make(chan os.Signal, 1)
	signal.Notify(sigCh, syscall.SIGINT, syscall.SIGTERM)
	go func() {
		<-sigCh
		log.Printf("signal received, draining…")
		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
		defer cancel()
		_ = httpServer.Shutdown(ctx)
	}()

	log.Printf("listening on :8080, redis=%s", addr)
	if err := httpServer.ListenAndServe(); err != nil && !errors.Is(err, http.ErrServerClosed) {
		log.Fatal(err)
	}
}

// waitForRedis pings Redis with backoff until it answers, then flips the
// readiness flag. /healthz returns 503 until this completes, which is
// what the grader's 60-second wait loop is designed to tolerate.
func (s *server) waitForRedis() {
	backoff := 250 * time.Millisecond
	for {
		ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
		err := s.rdb.Ping(ctx).Err()
		cancel()
		if err == nil {
			s.ready.Store(true)
			log.Printf("redis ready")
			return
		}
		time.Sleep(backoff)
		if backoff < 2*time.Second {
			backoff *= 2
		}
	}
}

func (s *server) healthz(w http.ResponseWriter, _ *http.Request) {
	if !s.ready.Load() {
		writeError(w, http.StatusServiceUnavailable, "not ready")
		return
	}
	writeJSON(w, http.StatusOK, map[string]string{"status": "ok"})
}

type shortenReq struct {
	URL string `json:"url"`
}

type shortenResp struct {
	Code     string `json:"code"`
	ShortURL string `json:"short_url"`
}

func (s *server) shorten(w http.ResponseWriter, r *http.Request) {
	raw, err := io.ReadAll(io.LimitReader(r.Body, 1<<14))
	if err != nil {
		writeError(w, http.StatusBadRequest, "invalid json body")
		return
	}
	var body shortenReq
	if err := json.Unmarshal(raw, &body); err != nil {
		writeError(w, http.StatusBadRequest, "invalid json body")
		return
	}
	url := strings.TrimSpace(body.URL)
	if url == "" {
		writeError(w, http.StatusBadRequest, "url is required")
		return
	}
	if !(strings.HasPrefix(url, "http://") || strings.HasPrefix(url, "https://")) {
		writeError(w, http.StatusBadRequest, "url must start with http:// or https://")
		return
	}

	code, err := s.mintCode(r.Context(), url)
	if err != nil {
		log.Printf("mint failed: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}
	writeJSON(w, http.StatusCreated, shortenResp{
		Code:     code,
		ShortURL: s.publicHost + "/" + code,
	})
}

// mintCode generates a fresh random code, persists it with SETNX, and
// retries a few times if it happens to collide with an existing key.
// With 62^7 ≈ 3.5e12 codes the collision probability is astronomically
// small for the scale we're testing at, but we still loop defensively.
func (s *server) mintCode(ctx context.Context, url string) (string, error) {
	for i := 0; i < maxMintTries; i++ {
		code := randomCode(codeLen)
		ok, err := s.rdb.SetNX(ctx, keyURLPrefix+code, url, 0).Result()
		if err != nil {
			return "", err
		}
		if ok {
			return code, nil
		}
	}
	return "", errors.New("code collision storm; retry")
}

func randomCode(n int) string {
	raw := make([]byte, n)
	if _, err := rand.Read(raw); err != nil {
		// crypto/rand failure is exotic. Don't 500 the request loop over it;
		// fall back to a low-quality time source. Modulo bias is irrelevant
		// for a short code.
		nano := time.Now().UnixNano()
		for i := range raw {
			raw[i] = byte(nano >> (uint(i) * 8))
		}
	}
	out := make([]byte, n)
	for i, b := range raw {
		out[i] = base62Alphabet[int(b)%62]
	}
	return string(out)
}

func (s *server) resolve(w http.ResponseWriter, r *http.Request) {
	code := r.PathValue("code")
	if !codeRegex.MatchString(code) {
		writeError(w, http.StatusNotFound, "not found")
		return
	}
	url, err := s.rdb.Get(r.Context(), keyURLPrefix+code).Result()
	if errors.Is(err, redis.Nil) {
		writeError(w, http.StatusNotFound, "not found")
		return
	}
	if err != nil {
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}

	// Fire-and-forget hit increment. The redirect path must not block on it
	// or we'd halve our RPS at the upper load stages.
	go func(c string) {
		ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
		defer cancel()
		_ = s.rdb.Incr(ctx, keyHitPrefix+c).Err()
	}(code)

	w.Header().Set("Location", url)
	w.WriteHeader(http.StatusFound)
}

func (s *server) metadata(w http.ResponseWriter, r *http.Request) {
	code := r.PathValue("code")
	if !codeRegex.MatchString(code) {
		writeError(w, http.StatusNotFound, "not found")
		return
	}
	res, err := s.rdb.MGet(r.Context(), keyURLPrefix+code, keyHitPrefix+code).Result()
	if err != nil {
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}
	if res[0] == nil {
		writeError(w, http.StatusNotFound, "not found")
		return
	}
	url, _ := res[0].(string)
	var hits int64
	if res[1] != nil {
		if hitsStr, ok := res[1].(string); ok {
			hits, _ = strconv.ParseInt(hitsStr, 10, 64)
		}
	}
	writeJSON(w, http.StatusOK, map[string]any{
		"code": code,
		"url":  url,
		"hits": hits,
	})
}

func getenv(k, def string) string {
	if v := os.Getenv(k); v != "" {
		return v
	}
	return def
}

func writeJSON(w http.ResponseWriter, status int, body any) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	_ = json.NewEncoder(w).Encode(body)
}

func writeError(w http.ResponseWriter, status int, msg string) {
	writeJSON(w, status, map[string]string{"error": msg})
}
